src/OpenTelemetry/Metrics/MeterProviderSdk.cs

```diff
@@ -412,12 +439,27 @@ internal Batch<Metric> Collect()
 
                     var indexSnapShot = Math.Min(this.metricIndex, MaxMetrics - 1);
                     var target = indexSnapShot + 1;
+                    int metricCountCurrentBatch = 0;
                     for (int i = 0; i < target; i++)
                     {
-                        this.metrics[i].SnapShot();
+                        var metric = this.metrics[i];
+                        if (metric != null)
+                        {
+                            if (metric.InstrumentDisposed)
+                            {
+                                metric.SnapShot();
```

Comment:
if the instrument was disposed *before* Collect, then only clear it in current Collect. Otherwise, it'll be cleaned up in the *next* collect only.

Reply:
What if the next Collect failed (e.g. timeout)?

Reply:
good point. The provider currently has no way of knowing that (only readers know that).